prover/src/trace/poly_table.rs

```diff
@@ -64,11 +72,27 @@ impl<E: FieldElement> TracePolyTable<E> {
         result
     }
 
-    /// Returns an out-of-domain evaluation frame constructed by evaluating trace polynomials
-    /// for all columns at points z and z * g, where g is the generator of the trace domain.
-    pub fn get_ood_frame(&self, z: E) -> Vec<Vec<E>> {
-        let g = E::from(E::BaseField::get_root_of_unity(self.poly_size().ilog2()));
-        vec![self.evaluate_at(z), self.evaluate_at(z * g)]
+    /// Returns an out-of-domain evaluation frame constructed by evaluating trace polynomials for
+    /// all columns at points z and z * g, where g is the generator of the trace
+    /// domain. Additionally, if the Lagrange kernel auxiliary column is present, we also evaluate
+    /// that column over the points: z, z * g, z * g^2, z * g^4, ..., z * g^(2^(v-1)), where v =
+    /// log(trace_len).
+    pub fn get_ood_frame(&self, z: E) -> OodFrameTraceStates<E> {
+        let log_trace_len = self.poly_size().ilog2();
+        let g = E::from(E::BaseField::get_root_of_unity(log_trace_len));
+        let current_frame = self.evaluate_at(z);
+        let next_frame = self.evaluate_at(z * g);
+
+        let lagrange_kernel_frame = self.lagrange_kernel_column_idx.map(|col_idx| {
+            let lagrange_kernel_col_poly = self.aux_segment_polys[0].get_column(col_idx);
+
+            LagrangeKernelEvaluationFrame::from_lagrange_kernel_column_poly(
+                lagrange_kernel_col_poly,
+                z,
+            )
+        });
+
+        OodFrameTraceStates::new(current_frame, next_frame, lagrange_kernel_frame)
```

Comment:
A few more thoughts about OOD frame structs.

It seems to me that we have two structs which contain basically identical data: `OodFrameTraceStates` (in the `air`  crate, but used primarily in the prover) and `TraceOodFrame` (used in the verifier). I wonder if we can unify these. For example, we could:

1. Replace the use of `TraceOodFrame` with `OodFrameTraceStates`. We might need to add some more methods to `OodFrameTraceStates` to do this - but I am not seeing any fundamental reasons why we can't do that.
2. Rename `OodFrameTraceStates` into `TraceOodFrame` (I like this name a little better).
3. Update the filed names as I mentioned in one of the previous comments. The struct can then look like this:

```Rust
pub struct TraceOodFrame <E: FieldElement> {
    current_row: Vec<E>,
    next_row: Vec<E>,
    lagrange_kernel_frame: Option<LagrangeKernelEvaluationFrame<E>>,
}
```

Then, `TracePolyTable::get_ood_frame()` method would return this struct and it would be passed to the prover channel.

One the verifier side, I think `OodFrame::parse()` could return this struct as well. The signature could look something like this:
```Rust
pub fn parse<E: FieldElement>(
    self,
    main_trace_width: usize,
    aux_trace_width: usize,
    num_evaluations: usize,
) -> Result<(TraceOodFrame<E>, Vec<E>), DeserializationError>
```

This should also nicely encapsulate the "interleaving" logic between current and next rows which is currently split between `OodFrame` and `TraceOodFrame`.

Reply:
I am all for this. Should we create a new issue though? This suggestion is pretty self-contained, and this PR is already very big.